src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncEventMap.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities;
+
+internal sealed class AsyncEventMap
+{
+    private readonly NonReentrantLock _guard = new();
```

Comment:
i would really like us to not use if this we can avoid it.  as this is a new type, please jsut use a SemaphoreSlim.